test/metabase/pulse/render/test_util.clj

```diff
@@ -33,3 +54,508 @@
+  [display-type viz-scenario]
+  (select-keys viz-scenario (valid-viz-scenarios display-type)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;    apply viz-scenarios
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn merge-viz-settings
+  "Merges the provided viz-settings into the `[:data :viz-settings]` path of the `card-and-data` map.
+  Notice that this does *not* merge settings into :visualization_settings of the card.
+
+  What is the difference between :visualization_settings in a card's data and [:data :viz-settings]?
+  The first is a key saved to the Card in the app-db, and does nothing until it is later used by the query processor's process-viz-settings middleware.
+  The latter is returned by the query processor under :data :viz-settings, if the process-viz-settinsg middleware
+  is run on the card's query. Since these util functions are mimicing a query result without running the processor,
+  we have to add the appropriate viz-settings in this location, as if the card's query was actually processed."
+  [card-and-data viz-settings]
+  (update-in card-and-data [:data :viz-settings] merge viz-settings))
+
+(defn- lisp-key
+  "Convert a keyword or string `k` from `snake_case` to `lisp-case`."
+  [k]
+  (if (keyword? k)
+    (keyword (lisp-key (name k)))
+    (str/replace k #"_" "-")))
+
+(defn- lisp-keys
+  "Convert the keys in a map from `snake_case` to `lisp-case`."
+  [m]
+  (u/recursive-map-keys lisp-key m))
+
+(defn- viz-namespaced-keys
+  [m]
+  (let [ns-key (fn [k]
+                 (keyword (str "metabase.shared.models.visualization-settings/" (name k))))]
+    (u/recursive-map-keys ns-key m)))
+
+(defn- merge-column-settings-for-data
+  "Merges the provided column-settings into the `[:data :column-settings]` path of the `card-and-data` map."
+  [card-and-data column-settings]
+  (update-in card-and-data [:data :viz-settings
+                            :metabase.shared.models.visualization-settings/column-settings]
+             (fn [a b] (merge-with merge a b)) (-> column-settings
+                                                   (update-vals lisp-keys)
+                                                   (update-vals viz-namespaced-keys))))
+
+(defn- merge-column-settings-for-card
+  "Merges the provided column-settings into the `[:card :visualization_settings :column_settings]` path of the `card-and-data` map."
+  [card-and-data column-settings]
+  (update-in card-and-data [:card :visualization_settings
+                            :column_settings]
+             (fn [a b] (merge-with merge a b)) (update-vals column-settings u/snake-keys)))
+
+(defn- make-settings-for-col
+  "Make the column-settings map for the given `col` the :card or :data `destination`.
+  This returns a map with one key and the settings.
+
+  The key will look like {:field-id 0} for :data
+  The key will be a json-stringified version of [\"ref\" [\"field\" 0 nil]] for :card."
+  [{field-ref :field_ref} settings destination]
+  (let [[_ id] field-ref]
```

Comment:
look into using the opposite side of `mb.viz/db->norm` from earlier: `mb.viz/norm->db`